### PR TITLE
Add pointer concept

### LINKED ISSUE #1643: [Sept Sprint] New exercise teaching pointers
This is part of our September Spring to get the Go Track in shape. Please read the [Overview issue](https://github.com/exercism/go/issues/1610) for more details. **This exercise is unclaimed. To claim it, please comment below.**

---

**Before building this exercise (or exercises), please read up on [the documentation](https://exercism.org/docs/building/tracks/concept-exercises).**

We need an exercise for the `pointers` and `nil` concepts. These concepts go well together so it makes sense to teach them in the same exercise.

The exercise should teach:

- What pointers are
- Pointer types for primitives
- Which types are already pointers
- Explain that go has no pointer arithmetic
- Referencing and dereferencing syntax
- `nil` as zero value for the pointer
- pitfalls when working with pointers (always check whether they are nil before dereferencing)

Keep in mind that the concept of pointers might be hard for people coming from languages that don't have explicit pointer types.

### PR DESCRIPTION
**Pull request by andrerfcsantos**
Fixes #1643

### CODE REVIEW

File: concepts/pointers/about.md
Comment (sudomateo) [inline concepts/pointers/about.md:105]: I'd personally recommend using another name for this struct type as the `Point` name is too similar to the pointer concept and may confuse readers. Perhaps a `Person` struct instead?

Comment (andrerfcsantos) [reply concepts/pointers/about.md:105]: That's a good point. After writing it down I also thought about that. I'll change it :)

Comment (sudomateo) [reply concepts/pointers/about.md:105]: Sorry to butt in! I was just poking around at the open pull requests. Appreciate your work here!

**Review by ErikSchierboom — COMMENTED** [review]
I'm really impressed with this PR! You did an excellent job. I've left some minor comments.

I don't really know anything about Go, so I'll leave it to others to review the Go parts of the code.

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:13]:
```suggestion
When we declare a variable, Go has to find a place in memory to store its value. This is largely abstracted from us — when we need to fetch the value stored in that piece of memory, we can just refer to it by the variable name.
```

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:21]:
```suggestion
For instance, when we write `a + 2`, we are effectively fetching the value stored in the memory associated with the variable `a` and adding 2 to it.
```

Comment (iHiD) [reply concepts/pointers/introduction.md:21]: I don't think this one needs to be split over multiple lines.

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:29]:
```suggestion
The piece of memory that is associated with `a` will now will now be storing the value `3`.
```

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:33]:
```suggestion
While variables allow us to refer to values in memory, sometimes it's useful to know the **memory address** to which the variable is pointing. **Pointers** hold the memory addresses of those values. A pointer declaration looks like this:
```

I've just tweaked this so that we show why they're called pointers (I always find if you can make the naming seem "natural" when teaching something, it sinks in better)

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:39]:
```suggestion
Here we declaring a variable `p` of type `*int`. The type `*int` means "pointer to int". In other words, we are saying that `p` is a pointer to int, which means that `p` will hold the memory address of an integer. The zero value of pointers is `nil` because a `nil` pointer holds no memory address.
```

Comment (iHiD) [reply concepts/pointers/introduction.md:39]: I felt those two sentences should be linked so it feels a little less like two seperate facts.

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:43]: I found this a little confusing, mainly because I'm not used to the word "operand". 

```suggestion
To find the memory address of the value of a variable, we can use the `&` operator. 
For example, if we want to find and store the member address of variable `a` in the pointer `p`, we can do the following:
```

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:50]:
```suggestion
p = &a // the variable 'p' will hold the memory address of 'a'
```

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:62]:
```suggestion
p = &a // the variable 'p' will hold the memory address of 'a'
```

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:55]:
```suggestion
When we have a pointer, we might want to know the value store in the memory address to which it points. We can do this using the `*` operator:
```

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:86]:
```suggestion
A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program panic at runtime!
```

Could we use a different word to "panic" here? Maybe "crash". I'm not sure someone coming from a different language (e.g. Ruby) would be familiar with the word, and so I have images in my mind of someone running around panicking, rather than the more technical usage :)

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:105]:
```suggestion
We could have also created a new `Person` and immediately stored a pointer to it:
```

File: concepts/pointers/introduction.md
Comment (andrerfcsantos) [reply concepts/pointers/introduction.md:43]: I was largely inspired by [Tour of Go: Pointers](https://tour.golang.org/moretypes/1) that also calls it operand.

But I agree the suggestion makes it a little more clear!

File: concepts/pointers/introduction.md
Comment (andrerfcsantos) [reply concepts/pointers/introduction.md:86]: In Go, a crash (in the sense of unhandled exception) is called a panic, that was the main reason for the use of panic there.

But I it doesn't hurt to make the introduction more inclusive and use a more general term.

Thanks for the suggestion.

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:3]: I think we need to give people a reason for why they're learning this before they work through such a complex topic. I've copied and shortened the bit at the bottom of document here. 

```suggestion
Like many other languages, Go has pointers. 
If you're new to pointers, they can feel a little mysterious but once you get used to them, they're quite straight-forward.
They're a crucial part of Go, so take some time to really understand them.

Before digging into the details, it's worth understanding the use of pointers. Pointers are a way to share memory with other parts of our program, which is useful for two major reasons:
1. When we have large amounts of data, making copies to pass between functions is very inefficient. 
  By passing the memory location of where the data is stored instead, we can dramatically reduce the resource-footprint of our programs.
2. By passing pointers between functions, we can access and modify the computer's memory directly, meaning that any changes made by one function are immediately visible to other parts of the program when the function ends.
```

File: concepts/pointers/introduction.md
Comment (iHiD) [inline concepts/pointers/introduction.md:148]: We can probably remove this from here now, as we have a summary version at the top. However, it'd be good to have this slightly more detailed version stay in the about.md

**Comment by iHiD** [comment]
I've reviewed the `introduction.md` - It's really great 🙂

**Comment by andrerfcsantos** [comment]
Thanks @iHiD and @ErikSchierboom for the extensive review!

These are great suggestions, really appreciate it. At a quick glance, I agree with all of them. I'll go through them in detail and submit a new patch.

File: exercises/concept/election-day/election_day_test.go
Comment (andrerfcsantos) [reply exercises/concept/election-day/election_day_test.go:14]: I like that suggestion a lot. I completely missed the `files.editor` key when reading through the docs.

Thanks for pointing that out.

**Comment by iHiD** [comment]
@andrerfcsantos I don't think I've ever seen Erik as impressed with a first-pass of a totally new exercise before 👏

File: exercises/concept/election-day/.meta/exemplar.go
Comment (andrerfcsantos) [reply exercises/concept/election-day/.meta/exemplar.go:39]: Added.

File: exercises/concept/election-day/.meta/exemplar.go
Comment (andrerfcsantos) [reply exercises/concept/election-day/.meta/exemplar.go:34]: Added.

File: exercises/concept/election-day/.meta/exemplar.go
Comment (andrerfcsantos) [reply exercises/concept/election-day/.meta/exemplar.go:3]: Added.

File: concepts/pointers/introduction.md
Comment (andrerfcsantos) [reply concepts/pointers/introduction.md]: Fixed.

File: concepts/pointers/introduction.md
Comment (andrerfcsantos) [reply concepts/pointers/introduction.md]: I like this change a lot!

I talked a bit with some people to know their troubles learning pointers before doing this exercise. One of the things they told me is that they had trouble learning pointers because they were not seeing why they were useful. That's why I made this little section, but I agree 100% that having this upfront is even better.

Change made.

**Comment by andrerfcsantos** [comment]
Changes applied. Ready for further review!

**Comment by iHiD** [comment]
Great! I probably won't have time to review again before GoBridge launches, so if you're happy with the changes you've applied, I am too :)

I guess we just need a review on the Go side of things and then we can merge :)

File: exercises/concept/election-day/.docs/instructions.md
Comment (andrerfcsantos) [reply exercises/concept/election-day/.docs/instructions.md]: Just checked and you are right. Fixed!

**Review by ErikSchierboom — APPROVED** [review]

**Comment by ErikSchierboom** [comment]
@iHiD Yep! Let's wait for a proper Go review, but I'll give it my approval to show the other reviewers I like this PR!

**Comment by bobahop** [comment]
Great work!

A couple of things.

For pointer to a slice, if the slice is appended to in a function it was passed into, the change will not be seen in the original, since the length and maybe capacity  will change, but the pointer is only for the address. So in that way a slice does not behave the same as a map. I only mention because they are grouped together in the explanation, and it mentions that map changes will be seen in the original.

The other issue is maybe to caveat against something like

`*p = &a`

I want to say I read somewhere that it's a common mistake for newer programmers to use the dereference operator when reassigning a pointer to a new address.

File: concepts/pointers/about.md
Comment (junedev) [inline concepts/pointers/about.md:88]: I think it is very important to show the actual error message here so people can recognize it later when it happens to them.
```suggestion
A note of caution however: always check if a pointer is not `nil` before dereferencing. Dereferencing a `nil` pointer will make the program crash at runtime!

```go
var p *int // p is nil initially
fmt.Println(*p)
// panic: runtime error: invalid memory address or nil pointer dereference
```

File: concepts/pointers/about.md
Comment (junedev) [inline concepts/pointers/about.md:145]: Suggestion how to address the slices issue mentioned here https://github.com/exercism/go/pull/1724#issuecomment-932351399
```suggestion

The same applies when changing an existing item in a slice. However, actions that return a new slice like `append` will **not** modify the slice outside of the function.
```

File: concepts/pointers/about.md
Comment (junedev) [inline concepts/pointers/about.md:148]: Suggesting to add an explanation of that pointer arithmetic actually is. Also rephrased the first bit a little.
```suggestion
Unlike other languages, Go does not support pointer arithmetic. This means you cannot go from a memory address of one value to the address of another by incrementing the pointer value. So snippets like the following are invalid:
```

File: concepts/pointers/introduction.md
Comment (junedev) [inline concepts/pointers/introduction.md:1]:
```suggestion
# Introduction
```

File: exercises/concept/election-day/.docs/instructions.md
Comment (junedev) [inline exercises/concept/election-day/.docs/instructions.md:9]:
```suggestion
count the votes was needed. The school needs your help building this new system.

## 1. Create a vote counter
```

File: exercises/concept/election-day/.meta/design.md
Comment (junedev) [inline exercises/concept/election-day/.meta/design.md:17]:
```suggestion
- Unsafe pointers
- Pointer receivers for methods (this is parts of the `methods` concept)
```

**Review by junedev — APPROVED** [review]
Great work! Really well done! :clap:  
Left some minor suggestions, if you like them they probably need to be applied in all the places.

File: concepts/pointers/about.md
Comment (moshegood) [reply concepts/pointers/about.md:150]: This is actually a common edge case in go.
```
func addOne(in []int) []int {
	return append(in, 1)
}

func main() {
	slice := []int{1, 2, 3, 4}
	slice = append(slice, 5)
	slice2 := append(slice, 6)
	fmt.Println(addOne(slice)) // prints 1,2,3,4,5,1
	fmt.Println(slice2)        // also prints 1,2,3,4,5,1
}
```

File: concepts/pointers/about.md
Comment (moshegood) [inline concepts/pointers/about.md:10]: s/computer's memory/single copy of the data/

File: concepts/pointers/about.md
Comment (moshegood) [inline concepts/pointers/about.md:132]: This is actually a potential edge case:
```
func addOne(in []int) []int {
	return append(in, 1)
}

func main() {
	slice := []int{1, 2, 3, 4}
	slice = append(slice, 5)
	slice2 := append(slice, 6)
	fmt.Println(addOne(slice)) // prints 1,2,3,4,5,1
	fmt.Println(slice2)        // also prints 1,2,3,4,5,1
}
```

File: concepts/pointers/about.md
Comment (moshegood) [inline concepts/pointers/about.md:151]: This is not necessarily true:
```
func addOne(in []int) []int {
	return append(in, 1)
}

func main() {
	slice := []int{1, 2, 3, 4}
	slice = append(slice, 5)
	slice2 := append(slice, 6)
	fmt.Println(addOne(slice)) // prints 1,2,3,4,5,1
	fmt.Println(slice2)        // also prints 1,2,3,4,5,1
}
```

Comment (andrerfcsantos) [reply concepts/pointers/about.md:151]: Maybe we can say "might not modify the slice" and link to a resource that explains this in more detail like [Arrays, slices (and strings): The mechanics of 'append'](https://go.dev/blog/slices).

**Comment by andrerfcsantos** [comment]
Made the suggested changes.

In addition I also made other improvements:

* In the pointer arithmetic example I declared the variables using the long form instead of the shorthand syntax. This is to 1) make the types of each variable more explicit 2) be more consistent with the rest of the examples
* I modified the map example slightly to increment a value on an existing map instead of adding a new key. This is just to be closer to what the student will be required to do in the exercise.

**Pull request merged**
